roofit/roofitcore/src/RooDataHist.cxx

```diff
@@ -599,7 +599,8 @@ void RooDataHist::_adjustBinning(RooRealVar &theirVar, const TAxis &axis,
     RooRealVar *ourVar, Int_t *offset)
 {
   if (!dynamic_cast<RooRealVar*>(static_cast<RooAbsArg*>(ourVar))) {
-    coutE(InputArguments) << "RooDataHist::adjustBinning(" << GetName() << ") ERROR: dimension " << ourVar->GetName() << " must be real" << endl ;
+    auto ourVarName = ourVar->GetName();
+    coutE(InputArguments) << "RooDataHist::adjustBinning(" << GetName() << ") ERROR: dimension " << (ourVarName != nullptr ? ourVarName : "undefined") << " must be real\n";
     assert(0) ;
```

Comment:
Bravo for wanting to improve that error message, but the entire thing is crazy.
- One shouldn't cast `RooRealVar` to `AbsArg` and dynamically back. That's useless.
- If somebody passes something that's not `RooRealVar` or derived into that function, it's time to throw `std::invalid_argument` (maybe better `logic_error`, because it's broken program logic if this happens).
- `assert(0)` is pretty much useless.

@guitargeek do you want to fix this (please?)? If @kgizdov has time to do this, we will happily accept that it, but don't feel obliged.

I suggest something like (github doesn't allow suggestions spanning over deleted lines):
```c++
if ( !dynamic_cast<RooRealVar*>(ourVar) ) throw std::logic_error("Wrong type passed to RooDataHist::_adjustBinning. Please report this issue.");
```

Reply:
I wasn't sure how much changes you guys were willing to accept. TBH, last night I almost rewrote half the class haha. I think that constructor has broken logic to start with. The comment insists that `numPDF==numCoef or numPDF==numCoef+1`, but it does not actually check for that. It also never validates in the for loop that after it ignores things, the number of PDFs and Coef matches in a desirable way.

There is also a point to be said about the following snippet:
```cpp
// RooAbsCollection.h
  Storage_t::size_type size() const {
    return _list.size();
  }
  ...
  inline Int_t getSize() const { 
    // Return the number of elements in the collection
    return _list.size();
  }
```
and :
```cpp
// RooArgList.
/// Return object at given index, or nullptr if index is out of range
  inline RooAbsArg* at(Int_t idx) const { 

    if (idx >= static_cast<Int_t>(_list.size()))
      return nullptr;

    return _list[idx];
  }
 ...
  /// Access element by index.
  RooAbsArg& operator[](Int_t idx) const {
    assert(0 <= idx && idx < static_cast<Int_t>(_list.size()));
    return *_list[idx];
  }
```
The interface has integer overflows all over.

Reply:
I know ... 🙁 
There's parts of the interface which have to stay as they are (for backward compatibility), and some that we just didn't get to fix yet. So, for all modernisations that can be done without breaking old code, I'm on board.

For those size issues you mention, I agree that you can accept `std::size_t` as argument. 👍
I'm not sure that you can return `std::size_t` from `getSize()`, though. It's worth a try. The function `size()` is meant as a way out, because it's conforming to what everybody expects, whereas `getSize()` should not be used in new code. If you want to try modernising, I'll be happy to have a look, but @guitargeek has to live with the fallout, so he should say if it's fine to go at it.

As to what you said about `numPDF`:
I'm not sure which class you are talking about, but if you can spare some time, we'll be happy about a helping hand and a PR. If the constructor doesn't check for the number of PDFs, please add those checks. It's better to refuse starting a fit, and then not needing to check while the fit is running.

Reply:
I was talking about `roofit/roofitcore/src/RooAddModel.cxx`, but I guess in the wrong comment thread hah.